pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -0,0 +1,257 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package nodeshutdown can watch for node level shutdown events and trigger graceful termination of pods running on the node prior to a system shutdown.
+package nodeshutdown
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/godbus/dbus/v5"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/kubelet/eviction"
+	"k8s.io/kubernetes/pkg/kubelet/nodeshutdown/systemd"
+	kubelettypes "k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/pkg/kubelet/util/format"
+)
+
+const (
+	nodeShutdownReason  = "Shutdown"
+	nodeShutdownMessage = "Node is shutting, evicting pods"
+)
+
+var getSystemDbus = systemDbus
+
+type dbusInhibiter interface {
+	CurrentInhibitDelay() (time.Duration, error)
+	InhibitShutdown() (systemd.InhibitLock, error)
+	ReleaseInhibitLock(lock systemd.InhibitLock) error
+	ReloadLogindConf() error
+	MonitorShutdown() (<-chan bool, error)
+	OverrideInhibitDelay(inhibitDelayMax time.Duration) error
+}
+
+// Manager has functions that can be used to interact with the Node Shutdown Manager.
+type Manager struct {
+	shutdownGracePeriodRequested    time.Duration
+	shutdownGracePeriodCriticalPods time.Duration
+
+	getPods eviction.ActivePodsFunc
+	killPod eviction.KillPodFunc
+
+	dbusCon     dbusInhibiter
+	inhibitLock systemd.InhibitLock
+
+	nodeShuttingDownMutex sync.Mutex
+	nodeShuttingDownNow   bool
+
+	clock clock.Clock
+}
+
+func systemDbus() (dbusInhibiter, error) {
+	bus, err := dbus.SystemBus()
+	if err != nil {
+		return nil, err
+	}
+	return &systemd.DBusCon{SystemBus: bus}, nil
+}
+
+// NewManager returns a new node shutdown manager.
+func NewManager(getPodsFunc eviction.ActivePodsFunc, killPodFunc eviction.KillPodFunc, shutdownGracePeriodRequested, shutdownGracePeriodCriticalPods time.Duration) *Manager {
+	return &Manager{
+		getPods:                         getPodsFunc,
+		killPod:                         killPodFunc,
+		shutdownGracePeriodRequested:    shutdownGracePeriodRequested,
+		shutdownGracePeriodCriticalPods: shutdownGracePeriodCriticalPods,
+		clock:                           clock.RealClock{},
+	}
+}
+
+// Start starts the node shutdown manager and will start watching the node for shutdown events.
+func (m *Manager) Start() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+	if m.shutdownGracePeriodRequested == 0 {
+		return nil
+	}
+
+	systemBus, err := getSystemDbus()
```

Comment:
I worry about the initial dial. Can systemd disconnect this socket and require another dial within the gofunction at line 148?

Reply:
Thanks for bringing it up. Based on the examples I found of using https://github.com/godbus/dbus, the system bus is connected to during startup and the connection is suppose to be persisted for the lifetime of the process. In my testing, I didn't see any cases where godbus library or systemd disconnected the socket.